src/mono/mono/mini/interp/transform.c

```diff
@@ -7576,6 +7576,11 @@ generate_code (TransformData *td, MonoMethod *method, MonoMethodHeader *header,
 			case CEE_UNUSED57: ves_abort(); break;
 #endif
 			case CEE_ENDFILTER:
+				if (td->sp - 1 != td->stack || td->sp->type != STACK_TYPE_I4) {
+					mono_error_set_generic_error (error, "System", "InvalidProgramException", "");
```

Comment:
`td->sp` points to the location where a new entry is allocated, so you are accessing junk here. You either access it with `td->sp [-1]`, or better decrement `td->sp` and also update accordingly the code below that adds the `MINT_ENDFILTER` instruction

Reply:
If we decrement `td->sp` and update accordingly the code that adds the `MINT_ENDFILTER` instruction, how `td-sp` is then rebased to the top of the stack in the next instruction?

Reply:
Probably here? https://github.com/dotnet/runtime/blob/a83c661bb8061b65404fb154b328ff1d3993d9ec/src/mono/mono/mini/interp/transform.c#L4595-L4602

Reply:
Each instruction consumes some values from the top of the stack and pushes some result back. `endfilter` consumes one value from the top of the stack so `td->sp` is technically decreased by 1. In practice, it doesn't really matter since `endifilter` is the last instruction of the bblock, so the next bblock resets the pointer according to its stack height https://github.com/dotnet/runtime/blob/main/src/mono/mono/mini/interp/transform.c#L4603

Reply:
Thanks for the explanation.